### PR TITLE
Performance - Cache full product objects 

### PR DESCRIPTION
**Pull request by claudiulodro**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

I was doing performance profiling around the Shop page and my profiling placed the product constructors near the top areas by time. This makes sense since loading products to display is basically all the Shop page does, so that’s not super interesting info on its own. The interesting part is that the profiling identified that there were many more requests to load data than there were products displayed on the Shop page. For example, we loaded product data 51 times on a Shop page with 16 products, and each product displayed in the Shop was loaded 3 times.

We do have plenty of caching around individual data pieces, and WordPress has caching around metadata. The main performance issue I found was that each time we load a product’s data 100+ functions run that set properties, sanitize data, apply filters, etc. We can improve performance quite a lot by not having to run these functions when fetching products we have already fetched.

When doing e.g. `get_post( $id )` multiple times for the same post, WordPress core [caches the full post objects](https://github.com/WordPress/WordPress/blob/7d17b53ff2882b09f7071bfa25a7d5972b3924a8/wp-includes/class-wp-post.php#L231-L255). A similar approach could help here, and in testing this approach improved performance on the Shop page dramatically.

I'm told something similar used to be in the product factory, but @franticpsyx noticed issues with it and it was removed. @franticpsyx, do you recall what the problem was in the past?

### Todo

- Product images on a lot of products disappear now for some reason. Other bugs may also be present. They should be investigated + fixed.
- Unit tests need to pass. Adding more/better clearing of the cached products should help with that.

### How to test the changes in this Pull Request:

These testing steps will not currently pass:
1. Verify product-based WooCommerce functionality (Shop, single products, admin, etc.) works correctly.
2. Verify unit tests pass.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### Changelog entry

> Performance - Cache full products to improve performance when loading the same product multiple times.

### CODE REVIEW

**Comment by claudiulodro** [comment]
@rrennick will be taking over this investigation. 🏆

**Comment by franticpsyx** [comment]
> do you recall what the problem was in the past

I tried to find the discussion we had with @mikejolley but didn't find anything in Slack or here, only this: https://github.com/woocommerce/woocommerce/commit/709d314ca738977ba7e289ea768002b908c148ea#r20124993

@claudiulodro said:

> The main performance issue I found was that each time we load a product’s data 100+ functions run that set properties, sanitize data, **apply filters**, etc.

Here's what I remember:

If you have any filters that run during object instantiation, then the initial object state will depend on external variables that can't be accounted for. So there's a very good chance we might cache something that's initialized in an unexpected way.

Such filters:

* might be run in the product type class constructor; or
* otherwise affect what's read from the DB in the datastore.

For example I think that back then, Variable product [price data](https://github.com/woocommerce/woocommerce/blob/master/includes/data-stores/class-wc-product-variable-data-store-cpt.php#L249) was read every time an instance was created. This is no longer the case nowadays.

If we are absolutely certain that the initial object state can't be manipulated using filters, then a cache here would work great. If there are any filters, then there _might_ be issues.

With that said, looking at the implementation, I can see it's very different (a lot safer!) compared to what I remember from 2 years ago. 

I'd feel safer moving this to 4.0, but if we can add some infrastructure to help third parties overcome any potential issues easily, then the performance benefits might outweigh any potential risk. A thought I had back then was to add a param to `wc_get_product` that specifically fetches a (non-cached) instance. This would give third parties the ability to add a hot-fix in their code without having to invalidate the cache for all subsequent calls in order to make their code work.

**Comment by franticpsyx** [comment]
Another thought -- I have the impression that `wp_cache_set` stores a shallow clone of the object. If that's true, then there might be implications when third-party code modifies an object prop (all instances will be affected?).

**Comment by rrennick** [comment]
> I have the impression that `wp_cache_set` stores a shallow copy of whatever is fed to it.

```
$shallow = new stdClass();
$shallow->data = [ 'A', 'B', 'C' ];
wp_cache_set( 'shallowdata', $shallow );
$shallow->data[] = 'D';
$cached = wp_cache_get( 'shallowdata' );
print_r( $shallow );
print_r( $cached );
```

Produces the following

```
stdClass Object
--
  | (
  | [data] => Array
  | (
  | [0] => A
  | [1] => B
  | [2] => C
  | [3] => D
  | )
  |  
  | )
  | stdClass Object
  | (
  | [data] => Array
  | (
  | [0] => A
  | [1] => B
  | [2] => C
  | )
  |  
  | )
```

**Comment by rrennick** [comment]
I've updated the PR to bypass the product cache when constructing product classes. This passes unit tests in my local environment but I have not done any other testing with it. I'm not comfortable that this is a good solution & will continue to investigate.

**Review by claudiulodro — COMMENTED** [review]
I think the approach and the code looks good so far. The unit tests pass, so that's a good sign. :)

In my testing, this latest version still provides a nice boost:

```
With this branch and other performance improvements from master branch: 2096 milliseconds
With just master branch: 2241 milliseconds
```

Next step is probably to do some research/testing around how other extensions are using this and make sure the potential filter issue mentioned in the comments here won't be a problem, etc.

**Comment by rrennick** [comment]
@claudiulodro Thanks for the feedback.

**Comment by rrennick** [comment]
@claudiulodro Can you fill me in on what theme, widgets, etc. you had configured for testing this?

**Comment by claudiulodro** [comment]
@rrennick I used to following:

- 500 products on site, generated with wc-smooth-generator
- Twenty Twelve theme (since it doesn't interact with WC in any way)
- Widgets shouldn't matter too much as long as they remain the same during testing (unless you have e.g. the price filter widget active with an active price filter or something else that would impact performance dramatically).

To get my performance measurements, I took the average of 5 load times of the Shop page using Query Monitor.

**Comment by rrennick** [comment]
I did some further testing and added debug log entries for the 3 cache operations. I tested with Storefront, Twenty Seventeen, and Twenty Twelve. 

My store was set up to display 16 products. The only WC widgets in the side bar were the price filter & product categories. The price widget adds one query to retrieve the min & max prices for the products that match the query.

On the store page or product category page, with or without the price filter I did not have any cache hits. However, when viewing a single product there were cache hits:

```
15-Feb-2019 15:28:07 UTC] cached: 1637
[15-Feb-2019 15:28:07 UTC] cache hit: 1637
[15-Feb-2019 15:28:07 UTC] cache hit: 1637
[15-Feb-2019 15:28:07 UTC] cache hit: 1637
[15-Feb-2019 15:28:07 UTC] cache hit: 1637
[15-Feb-2019 15:28:07 UTC] cache hit: 1637
[15-Feb-2019 15:28:07 UTC] cached: 511
[15-Feb-2019 15:28:07 UTC] cached: 373
[15-Feb-2019 15:28:07 UTC] cached: 823
[15-Feb-2019 15:28:07 UTC] cached: 485
[15-Feb-2019 15:28:07 UTC] cache hit: 485
[15-Feb-2019 15:28:07 UTC] cache hit: 511
[15-Feb-2019 15:28:07 UTC] cache hit: 373
[15-Feb-2019 15:28:07 UTC] cache hit: 823
[15-Feb-2019 15:28:07 UTC] cache hit: 1637
```

Variable product:

```
[15-Feb-2019 15:29:52 UTC] cached: 3308
[15-Feb-2019 15:29:52 UTC] cache hit: 3308
[15-Feb-2019 15:29:52 UTC] cache hit: 3308
[15-Feb-2019 15:29:52 UTC] cache hit: 3308
[15-Feb-2019 15:29:52 UTC] cache hit: 3308
[15-Feb-2019 15:29:52 UTC] cache hit: 3308
[15-Feb-2019 15:29:52 UTC] cached: 513
[15-Feb-2019 15:29:52 UTC] cached: 690
[15-Feb-2019 15:29:52 UTC] cached: 734
[15-Feb-2019 15:29:52 UTC] cached: 511
[15-Feb-2019 15:29:52 UTC] cache hit: 690
[15-Feb-2019 15:29:52 UTC] cache hit: 513
[15-Feb-2019 15:29:52 UTC] cache hit: 734
[15-Feb-2019 15:29:52 UTC] cache hit: 511
[15-Feb-2019 15:29:52 UTC] cache hit: 3308
```

I'll update the PR for the phpcs sniffs then is should be ready for more extensive testing/review.

**Comment by rrennick** [comment]
> then there might be implications when third-party code modifies a product prop that contains an object (all instances will be affected?).

The data store has the following filters:

```
~/Sites/ron/wp-content/plugins/woocommerce
$ grep -n apply_filters includes/data-stores/class-wc-product-data-store-cpt.php
101:			apply_filters(
490:				$download->set_file( apply_filters( 'woocommerce_file_download_path', $value['file'], $product, $key ) );
1002:		return (int) apply_filters( 'woocommerce_get_product_id_by_sku', $id, $sku );
1167:		$related_product_query = (array) apply_filters( 'woocommerce_product_related_posts_query', $this->get_related_products_query( $cats_array, $tags_array, $exclude_ids, $limit + 10 ), $product_id, $args );
1406:		$custom_results = apply_filters( 'woocommerce_product_pre_search_products', false, $term, $type, $include_variations, $all_statuses, $limit );
1792:		return apply_filters( 'woocommerce_product_data_store_cpt_get_products_query', $wp_query_args, $query_vars, $this );
```

and the following hooks

```
~/Sites/ron/wp-content/plugins/woocommerce
$ grep -n do_action includes/data-stores/class-wc-product-data-store-cpt.php
137:			do_action( 'woocommerce_new_product', $id );
255:		do_action( 'woocommerce_update_product', $product->get_id() );
280:			do_action( 'woocommerce_before_delete_' . $post_type, $id );
283:			do_action( 'woocommerce_delete_' . $post_type, $id );
287:			do_action( 'woocommerce_trash_' . $post_type, $id );
635:				do_action( 'woocommerce_variation_set_stock', $product );
637:				do_action( 'woocommerce_product_set_stock', $product );
643:				do_action( 'woocommerce_variation_set_stock_status', $product->get_id(), $product->get_stock_status(), $product );
645:				do_action( 'woocommerce_product_set_stock_status', $product->get_id(), $product->get_stock_status(), $product );
650:		do_action( 'woocommerce_product_object_updated_props', $product, $this->updated_props );
725:				do_action( 'woocommerce_product_set_visibility', $product->get_id(), $product->get_catalog_visibility() );
800:				do_action( 'woocommerce_process_product_file_download_paths', $product->get_parent_id(), $product->get_id(), $downloads );
802:				do_action( 'woocommerce_process_product_file_download_paths', $product->get_id(), 0, $downloads );
826:			do_action( 'woocommerce_product_type_changed', $product, $old_type, $new_type );
```

The filters look like they would all be unaffected by the cache,. To test the caching on a product edit, I edited product id `4259` and changed the stock amount. The cache activity was

```
[15-Feb-2019 17:08:58 UTC] cleaned cache: 4259
[15-Feb-2019 17:08:58 UTC] cleaned cache: 4259
[15-Feb-2019 17:08:58 UTC] cleaned cache: 4259
[15-Feb-2019 17:08:58 UTC] cached: 4259
[15-Feb-2019 17:08:58 UTC] cleaned cache: 4259
[15-Feb-2019 17:08:58 UTC] cleaned cache: 4259
[15-Feb-2019 17:08:58 UTC] cleaned cache: 4259
[15-Feb-2019 17:08:58 UTC] cleaned cache: 4259
[15-Feb-2019 17:08:58 UTC] cleaned cache: 4259
[15-Feb-2019 17:08:58 UTC] cached: 4259
[15-Feb-2019 17:08:58 UTC] cleaned cache: 4259
[15-Feb-2019 17:08:58 UTC] cleaned cache: 4259
[15-Feb-2019 17:08:58 UTC] cleaned cache: 4259
[15-Feb-2019 17:08:58 UTC] cleaned cache: 4259
[15-Feb-2019 17:08:58 UTC] cleaned cache: 4259
[15-Feb-2019 17:08:58 UTC] cached: 4259
[15-Feb-2019 17:08:58 UTC] cache hit: 4259
[15-Feb-2019 17:08:58 UTC] cache hit: 4259
```

That's typical of what I would expect the cache activity to look like.

**Comment by rrennick** [comment]
@mikejolley this should be ready for review.

File: includes/data-stores/class-wc-product-data-store-cpt.php [hunk 1]
```diff
@@ -163,7 +163,6 @@ public function read( &$product ) {
 				'short_description' => $post_object->post_excerpt,
 				'parent_id'         => $post_object->post_parent,
 				'menu_order'        => $post_object->menu_order,
-				'post_password'     => $post_object->post_password,
```
Comment (mikejolley) [inline includes/data-stores/class-wc-product-data-store-cpt.php]: Mistake?

File: includes/data-stores/class-wc-product-data-store-cpt.php [hunk 2]
```diff
@@ -174,6 +173,7 @@ public function read( &$product ) {
 		$this->read_product_data( $product );
 		$this->read_extra_data( $product );
 		$product->set_object_read( true );
+		wp_cache_set( 'product-' . $product->get_id(), $product, 'products' );
```
Comment (mikejolley) [inline includes/data-stores/class-wc-product-data-store-cpt.php]: What if we kept the cache set in the product factory? i.e. construct it, cache it, return it. That keeps the new caching code in one location.

**Comment by mikejolley** [comment]
@rrennick This change now looks a lot like the code we had before. I managed to trace back where it was reverted:

https://github.com/woocommerce/woocommerce/pull/13698

My understanding is that when the full object is cached, only the ID gets cached. It's reconstructed again later. This would make the cache in factory useless as before(?)...

**Comment by rrennick** [comment]
> My understanding is that when the full object is cached, only the ID gets cached.

Based on my test https://github.com/woocommerce/woocommerce/pull/22675#issuecomment-461432627 the caching makes a complete copy of the object. There is a small amount of overhead in making the copy.

On the serialization comment: that is the case for serialized options retrieved from the DB. `get_option` unserializes those every time they are retrieved from cache. Since this isn't using `*_option` there isn't any serialization.

File: includes/data-stores/class-wc-product-data-store-cpt.php (hunk 2 above)
Comment (rrennick) [reply includes/data-stores/class-wc-product-data-store-cpt.php]: @mikejolley great suggestion :)

**Comment by mikejolley** [comment]
@rrennick what about persistent caching? Perhaps thats what was being refered to. cc @kloon

**Comment by rrennick** [comment]
> what about persistent caching?

Memcache does clone but no serialization https://plugins.trac.wordpress.org/browser/memcached-redux/trunk/object-cache.php#L134. If there are DB based object caches still in use then they would probably need to serialize/unserialize.

**Comment by rrennick** [comment]
@mikejolley How about we add a filter that could be used to disable the caching?

File: includes/data-stores/class-wc-product-data-store-cpt.php (hunk 1 above)
Comment (rrennick) [reply includes/data-stores/class-wc-product-data-store-cpt.php]: Good catch, thanks.

**Comment by rrennick** [comment]
@mikejolley I've updated the PR with changes from the discussion.

**Comment by franticpsyx** [comment]
@rrennick 

I'll be as to the point as possible -- I am very worried about https://github.com/woocommerce/woocommerce/pull/22675/commits/2046d743b23b657eb70b7b73f50a29a66c48a5ff 

This has been tried and turned down in the past. 

> The filters look like they would all be unaffected by the cache. To test the caching on a product edit, I edited product id `4259` and changed the stock amount.

Any filters that *set an initial object state* will make that initial state persist. Stock amount (or other prop) changes flush the cache and should not cause any issues.

Another comment on:

https://github.com/woocommerce/woocommerce/pull/22675#issuecomment-461432627

You will need to test with object properties that are objects themselves to uncover potential issues -- not arrays. To my understanding, if `$product->property` is an object that gets created on instantiation, then the current solution will share the `$product->property` object across all instances obtained from the cache. Most notably, the data store itself is an object, so I guess data store instances will be shared. If a data store is coded in a way that sets props on load, then these will persist.

Core types may not have any issues with this, but other product types may break.

@rrennick what was the motivation behind https://github.com/woocommerce/woocommerce/pull/22675/commits/2046d743b23b657eb70b7b73f50a29a66c48a5ff ? Better performance?

**Comment by franticpsyx** [comment]
Here's an example of a data store property that I'd expect to be shared among all product instances retrieved from the same cache: https://github.com/woocommerce/woocommerce/blob/master/includes/data-stores/class-wc-product-variable-data-store-cpt.php#L24

We could try to create a failing test as follows:

1. Add a `woocommerce_variation_prices` filter and let it apply a 50% discount to prices.
2. Call `wc_get_product` on a variable product and then `$product->get_variation_prices()`. Note down the values.
3. Destroy (remove) the filter you added in step 1.
4. Call `wc_get_product` for the same product, and again `$product->get_variation_prices()`.

With https://github.com/woocommerce/woocommerce/commit/2046d743b23b657eb70b7b73f50a29a66c48a5ff in place I expect that in step (4) you'll get back the same filtered set of prices that you got in step (2), but I might be totally wrong.

**Comment by franticpsyx** [comment]
Added failing test. Doesn't fail on `master`.

**Comment by mikejolley** [comment]
I agree this looks problematic. I'd rather we cached DB lookups in datastores than full objects.

**Comment by rrennick** [comment]
> With [2046d74](https://github.com/woocommerce/woocommerce/commit/2046d743b23b657eb70b7b73f50a29a66c48a5ff) in place I expect that in step (4) you'll get back the same filtered set of prices that you got in step (2), but I might be totally wrong.

Generally, isn't caching the original unfiltered value the desirable behaviour? While there are situations where you need to add/remove filters dynamically, most filters are added once and may be called multiple times to process the same source data.

**Comment by franticpsyx** [comment]
@rrennick apologies for not being clear.

The desirable behavior in the failing test is this:

The `->get_variation_prices` call while the filter is **active** must return a different set of data compared to the one you get when you call `->get_variation_prices` **on a new instance** while the filter is **inactive** -- hence the use of `assertNotEquals`.

The assertion in the test fails because both instances share the same data store.

> While there are situations where you need to add/remove filters dynamically, most filters are added once and may be called multiple times to process the same source data.

That's a good question! 

In an **ideal** world, every filterable data-store operation that takes place **while data is read** should be re-run differently if the environment (filters) changes. If we wanted to be 100% safe with this, then we'd need to throw all caching ideas out of the window.

In practice, it's probably a good tradeoff to cache whatever is read by data-stores from the DB to prevent these "read" operations from running multiple times in a single request (some people might still have a problem with this). However, we should avoid caching entire objects (data stores or other product props that might be objects). 

In short: Different product instances could share the same _raw data_ throughout a request, but they should never share _state_.

**Comment by rrennick** [comment]
> The assertion in the test fails because both instances share the same data store.

Got it. I was missing out on the common DS.

For now, can we leave this one open so I can come back to it in a bit?

**Comment by mikejolley** [comment]
@rrennick @franticpsyx Thanks for experimenting with this and confirming there are problems with 
this approach. I've had a chat with @kloon and we definetely don't feel this is an approach we can ship with 3.6.

Caching objects = bad. Caching reads from DB = good.

If the existing changes in 3.6 still leave duplicate queries around, we should definetely look to add more caches in 3.6/3.7 and beyond, but I'd actually prefer we start a new PR to work on that so this one is left "for the record" in case the idea is pitched in the future :) As I mentioned earlier, we attempted this in the past and looks like we had the same conclusions :)

Thanks all!
